### ISSUE TITLE
IINA cannot open a file named in UTF8 NFC from Finder

### ISSUE THREAD

**Issue by Nathaniel-Wu** [iina/iina#4308]
**System and IINA version:**

- macOS 12.6.4 and 13.3
- IINA 1.3.1 Build 133

**Expected behavior:**
From Finder, open a file named "Å.mkv" (UTF8 NFD) with IINA, it should play, then close the player;
Rename the file to "Å.mkv" (UTF8 NFC), open it with IINA from Finder again, it should play, then close the player;
Use menu bar 'File > Open' to open the file, it should play, then quit IINA.

**Actual behavior:**
From Finder, open a file named "Å.mkv" (UTF8 NFD) with IINA, it plays, then close the player;
Rename the file to "Å.mkv" (UTF8 NFC), open it with IINA from Finder again, it does not play, in fact, no player window is open at all;
Use menu bar 'File > Open' to open the file, it plays, then quit IINA.

<details>
<summary><b>Crash report:</b></summary>

```
```
</details>

<br>

<details>
<summary><b>mpv log:</b></summary>

Open the file named in UTF8 NFC from Finder happens at 18:47:16.572; open it from menu bar happens at 18:47:21.354.
```
18:46:55.530 [iina][d] IINA 1.3.1 Build 133
18:46:55.531 [iina][d] Copyright © 2017-2022
18:46:55.531 [iina][d] Collider LI, et al.
18:46:55.531 [iina][d] Released under GPLv3.
18:46:55.531 [iina][d] FFmpeg 4.4.2
18:46:55.531 [iina][d]   libavcodec 58.134.100
18:46:55.531 [iina][d]   libavformat 58.76.100
18:46:55.531 [iina][d]   libavutil 56.70.100
18:46:55.531 [iina][d]   libswscale 5.9.100
18:46:55.531 [iina][d] App will launch
18:46:55.606 [iina][d] App launched
18:46:55.607 [iina][d] Loading key bindings
18:46:55.620 [iina][d] Set key bindings (63 mappings)
18:46:55.684 [player0][d] Set path to /usr/local/bin:/Applications/IINA.app/Contents/MacOS:/usr/bin:/bin:/usr/sbin:/sbin
18:46:55.711 [iina][d] Using mpv 0.34.1
18:46:55.714 [player0][d] Track list changed
18:46:55.807 [player0][d] Open URL: file://~/Downloads/A%CC%8A.mkv
18:46:55.807 [player0][d] Opening ~/Downloads/Å.mkv in main window
18:46:55.876 [iina][d] Created OpenGL pixel format with [__C._CGLPixelFormatAttribute(rawValue: 5), __C._CGLPixelFormatAttribute(rawValue: 96), __C._CGLPixelFormatAttribute(rawValue: 99), __C._CGLPixelFormatAttribute(rawValue: 12800), __C._CGLPixelFormatAttribute(rawValue: 73), __C._CGLPixelFormatAttribute(rawValue: 0)]
18:46:55.883 [iina][d] Falling back to nominal display refresh rate: 59.99286444870247 from 0.0
18:46:55.887 [iina][d] window!.screen: Built-in Retina Display visible frame (0.0, 94.0, 2560.0, 1321.0)
18:46:55.887 [iina][d] NSScreen.main: Built-in Retina Display visible frame (0.0, 94.0, 2560.0, 1321.0)
18:46:55.888 [iina][d] NSScreen.screens[0]: Built-in Retina Display visible frame (0.0, 94.0, 2560.0, 1321.0)
18:46:55.888 [iina][d] NSScreen.screens[1]: LG HDR 4K visible frame (2560.0, -550.0, 1440.0, 2535.0)
18:46:55.888 [iina][d] MainWindowLastPosition (960.0, 574.0) matched: Built-in Retina Display visible frame (0.0, 94.0, 2560.0, 1321.0)
18:46:56.013 [player0][d] File started
18:46:56.013 [player0][d] Track list changed
18:46:56.022 [player0][d] Started auto load
18:46:56.022 [fmatcher][d] **Start matching
18:46:56.042 [fmatcher][d] Getting all media files...
18:46:56.042 [fmatcher][d] Got all media files, video=1, audio=0
18:46:56.042 [fmatcher][d] Getting all sub files...
18:46:56.043 [fmatcher][d] Searching subtitles from 4 directories...
18:46:56.044 [fmatcher][d] Got 0 subtitles
18:46:56.044 [fmatcher][d] Adding files to playlist
18:46:56.044 [fmatcher][d] Grouping video files...
18:46:56.044 [fgroup][d] Start grouping 1 files
18:46:56.044 [fmatcher][d] Finished with 1 groups
18:46:56.044 [fmatcher][d] Grouping sub files...
18:46:56.044 [fgroup][d] Start grouping 0 files
18:46:56.044 [fmatcher][d] Finished with 1 groups
18:46:56.044 [fmatcher][d] Matching video and sub series...
18:46:56.044 [fmatcher][d] Calculated editing distance
18:46:56.044 [fmatcher][d] Finished matching.
18:46:56.044 [fmatcher][d] Matching subs with matched series, prefixes=0...
18:46:56.044 [fmatcher][d] Matching for Å
18:46:56.044 [fmatcher][d] No matched sub for this file
18:46:56.044 [fmatcher][d] Force matching unmatched videos, video=1, sub=0...
18:46:56.044 [fmatcher][d] **Finished matching
18:46:56.193 [player0][d] File loaded
18:46:56.201 [player0][d] Track list changed
18:46:56.364 [player0][d] Getting thumbnails
18:46:56.365 [player0][d] Found thumbnail cache
18:46:56.365 [thumbcache][d] Reading thumbnail cache...
18:46:56.365 [thumbcache][d] Reading from ~/Library/Caches/com.colliderli.iina/thumb_cache/a84e82b042906e3c00733004ccd75fee
18:46:56.377 [thumbcache][d] Finished reading thumbnail cache, 45 in total
18:46:56.394 [player0][d] Track list changed
18:46:56.662 [player0][d] Playback restarted
18:46:58.436 [player0][d] Write watch later config
18:46:58.437 [player0][d] Stopping playback
18:46:58.497 [player0][d] Playback has stopped
18:46:58.498 [player0][d] Track list changed
18:47:16.572 [iina][d] Handle reopen
18:47:21.354 [iina][d] Menu - Open file
18:47:24.659 [player0][d] Open URL: file://~/Downloads/%C3%85.mkv
18:47:24.660 [player0][d] Opening ~/Downloads/Å.mkv in main window
18:47:24.660 [iina][d] window!.screen: Built-in Retina Display visible frame (0.0, 94.0, 2560.0, 1321.0)
18:47:24.660 [iina][d] NSScreen.main: Built-in Retina Display visible frame (0.0, 94.0, 2560.0, 1321.0)
18:47:24.660 [iina][d] NSScreen.screens[0]: Built-in Retina Display visible frame (0.0, 94.0, 2560.0, 1321.0)
18:47:24.660 [iina][d] NSScreen.screens[1]: LG HDR 4K visible frame (2560.0, -550.0, 1440.0, 2535.0)
18:47:24.660 [iina][d] MainWindowLastPosition (960.0, 574.0) matched: Built-in Retina Display visible frame (0.0, 94.0, 2560.0, 1321.0)
18:47:24.663 [player0][d] File started
18:47:24.663 [player0][d] Track list changed
18:47:24.672 [player0][d] Started auto load
18:47:24.672 [fmatcher][d] **Start matching
18:47:24.673 [fmatcher][d] Getting all media files...
18:47:24.673 [fmatcher][d] Got all media files, video=1, audio=0
18:47:24.673 [fmatcher][d] Getting all sub files...
18:47:24.674 [fmatcher][d] Searching subtitles from 4 directories...
18:47:24.675 [fmatcher][d] Got 0 subtitles
18:47:24.675 [fmatcher][d] Adding files to playlist
18:47:24.675 [fmatcher][d] Grouping video files...
18:47:24.675 [fgroup][d] Start grouping 1 files
18:47:24.675 [fmatcher][d] Finished with 1 groups
18:47:24.675 [fmatcher][d] Grouping sub files...
18:47:24.675 [fgroup][d] Start grouping 0 files
18:47:24.675 [fmatcher][d] Finished with 1 groups
18:47:24.675 [fmatcher][d] Matching video and sub series...
18:47:24.675 [fmatcher][d] Calculated editing distance
18:47:24.675 [fmatcher][d] Finished matching.
18:47:24.675 [fmatcher][d] Matching subs with matched series, prefixes=0...
18:47:24.675 [fmatcher][d] Matching for Å
18:47:24.675 [fmatcher][d] No matched sub for this file
18:47:24.675 [fmatcher][d] Force matching unmatched videos, video=1, sub=0...
18:47:24.675 [fmatcher][d] **Finished matching
18:47:24.704 [player0][d] File loaded
18:47:24.713 [player0][d] Getting thumbnails
18:47:24.714 [player0][d] Found thumbnail cache
18:47:24.714 [player0][d] Track list changed
18:47:24.714 [thumbcache][d] Reading thumbnail cache...
18:47:24.714 [thumbcache][d] Reading from ~/Library/Caches/com.colliderli.iina/thumb_cache/a84e82b042906e3c00733004ccd75fee
18:47:24.722 [thumbcache][d] Finished reading thumbnail cache, 45 in total
18:47:24.751 [player0][d] Track list changed
18:47:25.078 [player0][d] Playback restarted
18:47:27.105 [hdr][d] HDR primaries and gamma not available
18:47:27.119 [hdr][d] HDR primaries and gamma not available
18:47:27.132 [player0][d] Player has shutdown (initiated by mpv)
18:47:27.134 [iina][d] App should terminate
18:47:27.134 [iina][d] Disabling all menus
18:47:27.135 [iina][d] Closing all windows
18:47:27.137 [iina][d] All players have shutdown; proceeding with application termination
18:47:27.154 [iina][d] App will terminate
```
</details>

<br>

**Steps to reproduce:**
In Finder, name any video file to "Å" or any other Unicode text that differs in NFC and NFD, Finder will use UTF8 NFD to save the file name (the extension name and video format does not matter);
From Finder, open this video file with IINA, then close the player;
Rename the file in UTF8 NFC with `convmv -f utf8 -t utf8 --nfc --notest /path/to/file` (you may need to install it via Homebrew), open it with IINA from Finder again, it will not play;
Use menu bar 'File > Open' to open the file, it should play, then quit IINA.

- [x] MPV does not have this problem.

**How often does this happen?**
Always.

**Comment by low-batt** [comment]
Reproduced for me with IINA, mpv and VLC…

The following testing was performed on a MacBookPro18,2 running macOS Ventura 13.3.

This is from `iina.log` in a debug version of IINA where I added a log message in the [application](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428612-application) method called by macOS:
```text
4:04:13.062 [iina][d] App will launch
14:04:13.094 [iina][d] AppDelegate.application filename /Users/low-batt/Documents/builds/iina-official/issue-4308/Å.mp4
14:04:13.095 [iina][d] App launched
```

I installed `convmv` using [Homebrew](https://brew.sh/) and used it to rename the file:
```bash
low-batt@gag issue-4308$ convmv -f utf8 -t utf8 --nfc --notest Å.mp4
mv "./Å.mp4"	"./Å.mp4"
Ready! I converted 1 files in 0 seconds.
low-batt@gag issue-4308$ 
```

I then tried opening the renamed file using [Finder](https://support.apple.com/en-us/HT201732). IINA did not open the file, as reported above. The `iina.log` showed:
```text
14:06:41.266 [iina][d] App will launch
14:06:41.300 [iina][d] App launched
```

Attempting to open the file by right clicking on the file in `Finder` and selecting [mpv](https://mpv.io/) or [VLC](https://www.videolan.org/vlc/) I see similar behavior to IINA where both of those players behave as if they were started without specifying a file.

Looking at `man convmv` there is a discussion about Apple enforcing NFD under previous file system formats and relenting with the introduction of [APFS](https://en.wikipedia.org/wiki/Apple_File_System):
>  Apple, with the introduction of APFS in macOS 10.3, gave up to impose NFD on user space. But once you enforced NFD there is no easy way back without breaking existing applications. So they had to make APFS normalization-insensitive, that means a file can be created in NFC or NFD in the filesystem and it can be accessed with both forms also. Under the hood they store hashes of the normalized form of the filename to provide normalization insensitivity. Sounds like a great idea? Let's see: If you readddir a directory, you will get back the files in the the normalization form that was used when those files were created. If you stat a file in NFC or in NFD form you will get back whatever normalization form you used in the stat call. So user space applications can't expect that a file that can be stat'ed and accessed successfully, is also part of directory listings because the returned normalization form is faked to match what the user asked for. Theoretically also user space will have to normalize strings all the time. This is the same problem as for the case insensitivity of filenames before, which still breaks many user space applications. Just that the latter one was much more obvious to spot and to implement than this thing. So long, and thanks for all the fish.

Maybe `Finder` still needs to be updated to handle NFC?

As other players show this same problem and debugging points to `Finder` not passing the filename to open I'm labeling this as an upstream problem in macOS.

**Comment by Nathaniel-Wu** [comment]
It does seem to be an upstream problem, I just noticed my other applications now also have this problem with similar files, must be the new OS update.

**Comment by low-batt** [comment]
Thank you for letting us know you are seeing this same problem with other applications.

After updating to macOS 13.3 I experienced a window manager hang similar to a problem seen during the Ventura beta. Sure hope Apple is working on these regressions.

**Comment by StreetWalker92** [comment]
I don't know if this issue is totally related with the above, but it behaves almost identical. When I try to open a video file with a 'ñ' character, "España.mp4" for example, IINA will open but just the app, not the video. I tried adding a 'ñ' character to other files, like images, but their corresponding apps launch it correctly.

I'm running IINA 1.3.1 and macOS 13.3.

**Comment by low-batt** [comment]
@StreetWalker92 Sounds like this same issue.

I copied "España.mp4" from the post above and named a file that. Double clicking in `Finder` opened IINA and played the file. I then did this:
```bash
low-batt@gag issue-4308$ convmv -f utf8 -t utf8 --nfc --notest España.mp4 
mv "./España.mp4"	"./España.mp4"
Ready! I converted 1 files in 0 seconds.
low-batt@gag issue-4308$ 
```

And now double clicking starts IINA, but IINA does not start playing the file.

I then created a text file named "España.txt". Double clicking on that does not open the file in `TextEdit`.

I'm still looking around for a way to display the character encoding used in a filename.

The macOS behavior is sensitive to the file system format. Is the file you are testing on an internal disk, or a network volume?

**Comment by StreetWalker92** [comment]
Oh, that's good to know. The file is on my internal SSD, which is in APFS format. I copied the file to an external USB HDD in NTFS format and it plays fine. Which is interesting tough, is that I returned to my original SSD file and it plays fine now! Then I created a copy and name it exactly as the original but with a '2' at the end, and doesn't played fine. Odd to say the least.

**Comment by low-batt** [comment]
In a post above @Nathaniel-Wu mentioned:
> must be the new OS update.

I'm assuming that means this was working in the past and something has changed recently in macOS to break it.

Opening the file using [Terminal](https://support.apple.com/guide/terminal/welcome/mac):
```bash
low-batt@gag issue-4308$ open Å.mp4
```
Also exhibits the same problem.

**Comment by StreetWalker92** [comment]
Yes, I've never had issues playing files with IINA, and yes, this is after the 13.3 update. Given the facts, I don't think there's a problem with IINA. Thank you for your patiente and your continuous work 🙏

**Comment by low-batt** [comment]
I view this as a valid IINA problem and will be leaving this issue open until this is working again. Even though it seems to be an "upstream" problem in macOS that IINA developers can't fix, it is still a problem for IINA users. Glad both of you reported this odd unexpected behavior. Now if another user encounters this they might find this issue and not have to keep searching the net for an answer as to what is going wrong.

On GitHub there is more than one view as to what an issue is for. GitHub itself views an issue as a mechanism for tracking work for developers to do. GitHub automation closes an issue once a fix is merged. You will see IINA developers reopening issues GitHub has closed (when we remember) as the IINA project takes the view that it is still an open problem for users until a new release containing the fix is available. In other words IINA treats issues as a list of known problems with the current release of IINA.

**Comment by low-batt** [comment]
Well… It appears the `Finder` problem extends well beyond what was discussed above. Today when I click on text files that do not have special characters in the filename `Finder` acts as if the file was opened by `TextEdit`. The `Date Last Opened` is updated. But a `TextEdit` window does not appear. Checking `TextEdit` the file is not listed in `Open Recent`.

I'm starting to think macOS Ventura 13.3 has been enhanced with brand new advanced creative exhilarating regressions.

Poking around the net I found the post [macOS Ventura 13.3 alters expected behavior for Finder’s Open With functionality for macOS installer packages](https://derflounder.wordpress.com/2023/04/01/macos-ventura-13-3-alters-expected-behavior-for-finders-open-as-functionality-for-macos-installer-packages/). That post is not about the above problems, but it confirms macOS 13.3 contains changes in this area.

**Comment by low-batt** [comment]
I've "upgraded" my development MacBook Pro to macOS 13.3.1. I just finished testing to see if Apple has fixed this significant regression. I doubled clicked on a test file named "Å.txt" in `Finder` and `TextEdit` did not open the file. Seems this regression is still present in macOS 13.3.1.

**Comment by low-batt** [comment]
This is such a bad regression that Apple mentioned it in release notes. From [macOS Ventura 13.4 Beta 2 Release Notes](https://developer.apple.com/documentation/macos-release-notes/macos-13_4-release-notes):
> Fixes a regression in macOS Ventura 13.3 where a security check causes bookmark resolution to fail when the path contains Unicode characters stored with composed normalization. As an example, this prevented files in Finder from opening when double-clicked. (107550080)

Seem a fix is coming in macOS Ventura 13.4. Hopefully without any new exciting regressions.

**Comment by StreetWalker92** [comment]
I have installed macOS Ventura 13.4, which have been released right now. I can confirm that this issue is gone.

**Comment by low-batt** [comment]
Thanks for confirming macOS 13.4 fixes this issue. Now that macOS is working again (at least with respects to this problem) I'm closing this issue.

**Comment by Nathaniel-Wu** [comment]
Tried with the new macOS 12.6.6 as well, and it also works again

**Comment by StreetWalker92** [comment]
Well, oddly enough nearly 3 years have been passed and right now I'm encountering with the exact same issue on the exact video file. If I right-click -> Open with another player it works fine. But 3 years ago this was a problem with macOS itself, I started noticing this problem with 26.6.2. I updated to 27 Beta 6 and it is the same. I think I need to find a workaround to rename the file to not encounter this issue ever again.

**Comment by low-batt** [comment]
@StreetWalker92 I just upgraded to macOS 26.6.2. I still had the files I used to reproduce the problem. I tried double clicking on them and IINA opened and played the files. Not sure why it did not reproduce for me.

We do have reports of bad problems happening under the macOS 27 beta. Problems with audio not starting and causing crashes.

**Comment by StreetWalker92** [comment]
@low-batt Hey, thanks for reaching out. I recorded the issue and attached the IINA logs.

Video: https://we.tl/t-3EWNjV6tmPm3i3TH

What you will see on the video:
1. I made a copy of the file to my desktop.
2. The copy on my desktop plays perfectly fine with IINA.
3. The original file on my hard drive plays fine if I right-click -> Open with VLC
4. If I double-click on the original file, IINA seems to "freeze" and Finder enters the state of "Not responding". I have to force shut down my computer to recover from this.

I know the file name doesn´t have any strange character, but 3 years ago neither. It is in the same route/folder than 3 years ago and this problem was fixed by Ventura 13.4 and I never experienced this problem again until 26.6.2. I uptated to 27 Beta 6 but it is the same.

I attached the IINA logs as well. I will be very grateful if you can shed some light into this. Thank you.

[2026-08-24-16-23-51_sTIYu0.zip](https://github.com/user-attachments/files/31379625/2026-08-24-16-23-51_sTIYu0.zip)

**Comment by low-batt** [comment]
Sorry I'm slow to respond. Totally overloaded…

The screen recording was especially helpful. This is not the old problem reported in this issue. The `mpv.log` file is zero length. Likely that log contains some important information about this problem.

Seems like there are multiple problems…

That Finder is showing the [spinning beach ball of death](https://en.wikipedia.org/wiki/Spinning_pinwheel) looks like a macOS 27 defect to me.

That IINA's `Welcome Window` is not showing any files is disturbing. In IINA's settings on the `General` tab in the `History` section you have `Enable "Open Recent" menu` and `Track all played files in "Open Recent menu"` both checked?

The list of recently opened files is owned / managed by Apple's [AppKit](https://developer.apple.com/documentation/appkit) Framework. Are you seeing that list getting erased? That can happen if you run an IINA Nightly build which is unsigned. Apple made a change where if the signature of the app accessing the list changes, AppKit **ERASES THE USER'S DATA**. No idea why Apple thinks it is fine to erase the user's data. I mention this because if something goes wrong with this system it might incorrectly erase the list. I'm thinking that might be what is happening. Or it could be that the macOS daemon that manages this list has locked up.

Sounded like the audio is messed up under IINA? The `mpv.log` might have some information about that. There is a known problem with the `Core Audio` driver under macOS 27. See IINA issue  #6274. The currently thinking is that this is a problem in `libmpv`, rather than macOS 27, but it is only reproducible under macOS 27, so a macOS defect cannot be fully ruled out yet.  Issue #6279 is definitely a `libmpv` problem. IINA is waiting for the mpv project to fix these problems. 

_However_ this is a [.ts](https://en.wikipedia.org/wiki/.ts) file (MPEG-TS). There are known problems with mpv and `.ts` files, although I've not seen anything like this. Usually those problems have to do with seeking. VLC uses its own code to process such files and handles them better than mpv which uses FFmpeg. At least that is my understanding.

There is also this in the IINA log:
```text
16:24:16.175 [iina][w] Timed out waiting for players to stop and shutdown
16:24:16.175 [iina][w] Player 0 failed to stop
16:24:16.175 [iina][w] Not waiting for players to shutdown; proceeding with application termination
16:24:16.176 [iina][w] Proceeding with application termination due to time out
```

That should not be causing any of these other problems, but that shouldn't be happening.

Probably should be discussing this in a new issue…

**Comment by StreetWalker92** [comment]
@low-batt Hi. Thanks for reaching out again :)

- The mvp.log file is zero lenght because it is zero lenght indeed. If I reproduce any other file the mvp.log shows the intended information, but not in this case. I attached it anyway to show you that I didn't forget to upload the mvp log.

- The IINA's Welcome Menu not showing any recent file is totally my fault. I always disable options like that on my apps and they are disabled in IINA as well. 

- I'm using IINA's release version.

- The audio seems fine to me, maybe it sounded weird 'cause of the video compression. But when I play the copy on my desktop is totally fine to my ears.

If you need me to open an issue for this matter, let me know. I'll do it in a second. Thank you.

**Comment by low-batt** [comment]
I've entered issue #6302 for this. See if I reported it correctly.

**Comment by low-batt** [comment]
Late here and I've already proven today that I have run out of brain power…

On this:
> The mvp.log file is zero lenght because it is zero lenght indeed. If I reproduce any other file the mvp.log shows the intended information, but not in this case. I attached it anyway to show you that I didn't forget to upload the mvp log.

That is interesting. I will look into that.

On this:
> The IINA's Welcome Menu not showing any recent file is totally my fault. I always disable options like that on my apps and they are disabled in IINA as well.

Glad that is a non-issue.

On this:
> I'm using IINA's release version.

More evidence I'm overloaded and burned out.

On this:
> The audio seems fine to me, maybe it sounded weird 'cause of the video compression. But when I play the copy on my desktop is totally fine to my ears.

I was completely miss-interpreting what was happening. Clearly I need sleep. I will think on this problem and respond when I am coherent.
